cli/src/host_perf_check.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright 2017-2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::error::PerfCheckError;
+use log::info;
+use nix::unistd;
+use polkadot_node_core_pvf::sp_maybe_compressed_blob;
+use std::{
+	fs::{self, OpenOptions},
+	io::{self, Read, Write},
+	path::{Path, PathBuf},
+	time::{Duration, Instant},
+};
+
+fn check_passed_file_path() -> io::Result<PathBuf> {
+	let home_dir =
+		std::env::var("HOME").map_err(|err| io::Error::new(io::ErrorKind::Other, err))?;
+	let path = Path::new(&home_dir).join(".polkadot");
+
+	if let Err(err) = fs::create_dir(&path) {
+		if err.kind() != io::ErrorKind::AlreadyExists {
+			return Err(err)
+		}
+	}
+
+	Ok(path.join("perf_check_passed"))
+}
+
+fn is_perf_check_done(path: &Path) -> io::Result<bool> {
+	let host_name_max_len = unistd::SysconfVar::HOST_NAME_MAX as usize;
+
+	let mut host_name = vec![0u8; host_name_max_len];
+	let mut buf = host_name.clone();
+	unistd::gethostname(&mut host_name).map_err(|err| io::Error::new(io::ErrorKind::Other, err))?;
+
+	let file = match fs::File::open(path) {
+		Ok(file) => file,
+		Err(err) if err.kind() == io::ErrorKind::NotFound => return Ok(false),
+		Err(err) => return Err(err),
+	};
+	let mut reader = io::BufReader::new(file);
+
+	reader.read_exact(&mut buf)?;
+
+	Ok(host_name == buf)
+}
+
+fn save_check_passed_file(path: &Path) -> io::Result<()> {
+	let host_name_max_len = unistd::SysconfVar::HOST_NAME_MAX as usize;
+	let mut host_name = vec![0u8; host_name_max_len];
+	unistd::gethostname(&mut host_name).map_err(|err| io::Error::new(io::ErrorKind::Other, err))?;
+
+	let mut file = OpenOptions::new().truncate(true).create(true).write(true).open(path)?;
+
+	file.write(&host_name)?;
+
+	Ok(())
+}
+
+pub fn host_perf_check() -> Result<(), PerfCheckError> {
+	const PERF_CHECK_TIME_LIMIT: Duration = Duration::from_secs(20);
+	const CODE_SIZE_LIMIT: usize = 1024usize.pow(3);
+	const WASM_CODE: &[u8] = include_bytes!(
+		"../../target/release/wbuild/kusama-runtime/kusama_runtime.compact.compressed.wasm"
```

Comment:
This will also need to be changed.

The runtime itself already expose the `WASM_CODE` constant.

Reply:
Could you give a hint on how I can access it? I thought it's not present in the binary.

Reply:
`kusama_runtime::WASM_BINARY`

Reply:
Hm, during compile time it's `None` for me, but `Some` during runtime, which makes it impossible to put the code into `const` variable.
Any way around this?